crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1207,15 +1207,41 @@ impl<'db> TypeInferenceBuilder<'db> {
                     _ => false,
                 }
             }
-            let is_overload = function.decorator_list.iter().any(|decorator| {
+
+            let is_overload_or_abstract = function.decorator_list.iter().any(|decorator| {
                 let decorator_type = self.file_expression_type(&decorator.expression);
 
-                decorator_type
-                    .into_function_literal()
-                    .is_some_and(|f| f.is_known(self.db(), KnownFunction::Overload))
+                match decorator_type {
+                    Type::FunctionLiteral(function) => matches!(
+                        function.known(self.db()),
+                        Some(KnownFunction::Overload | KnownFunction::AbstractMethod)
+                    ),
+                    _ => false,
+                }
             });
-            // TODO: Protocol / abstract methods can have empty bodies
-            if (self.in_stub() || is_overload)
+
+            let class_inherits_protocol_directly = (|| -> bool {
+                let current_scope = self.scope().scope(self.db());
+                let Some(parent_scope_id) = current_scope.parent() else {
+                    return false;
+                };
+                let parent_scope = parent_scope_id
+                    .to_scope_id(self.db(), self.file())
+                    .scope(self.db());
```

Comment:
Nit: We don't need to get the `ScopeId` here because we know that all scopes are from the same file (this is slightly faster)
```suggestion
                let current_scope_id = self.scope().file_scope_id(self.db());
                let current_scope = self.index.scope(current_scope_id);

                let Some(parent_scope_id) = current_scope.parent() else {
                    return false;
                };
                let parent_scope = self.index.scope(parent_scope_id);
```